### DISCUSSION TITLE
Add autowiring for laravel container

### DISCUSSION THREAD

**Discussion by asanikovich** [laravel/framework#46638] [Ideas]
Based on symfony container autowiring https://symfony.com/doc/current/service_container/autowiring.html#working-with-interfaces I want to offer to add autowiring for laravel container.
Example:
We have Interface
```php
namespace App\Contract;

interface TransformerInterface
{
    public function transform(string $value): string;
}
```

And implementation
```php
namespace App\Util;

class Rot13Transformer implements TransformerInterface
{
    // ...
}
```

Usage in code
```php
namespace App\Http\Controllers;

use App\Contract\TransformerInterface;

class Controller
{
    public function __construct(TransformerInterface $transformer)
    {
    }
}

```

If we try to run now, we will have the error:
  ```
 Illuminate\Contracts\Container\BindingResolutionException

  Target [App\Http\Controllers\Controller] is not instantiable while building [App\Contract\TransformerInterface].
```
To fix it we need bind interface with class explicitly.

When autowiring is enabled and if only one class exists that implements an interface, we can automatically create bind them:
`for any usage of TransformerInterface provide Rot13Transformer.`

What do you think, community?

**Comment by Angel5a** [comment]
Personally I'm not a fun of such magic behavior. In my opinion interface should be explicitly bind to implementation. I don't like MailSender from some third party package would be called just because it exists and I miss to bind interface MailSenderInterface to some RiseAnErrorSender. I'd riser get an instantiation error on resolving MailSenderInterface and choose appropriate implementation by my self.
